test/subscription-client.js

```diff
@@ -0,0 +1,200 @@
+    t.deepEqual(data, {
+      topic: 'test-service_1',
+      payload: {
+        foo: 'bar'
+      }
+    })
+    client.close()
+    server.close()
+    t.end()
+  })
+})
+
+test('subscription client calls the publish method with null after GQL_COMPLETE type payload received', (t) => {
+  const server = new WS.Server({ port: 8889 })
+  server.on('connection', function connection (ws) {
+    ws.on('message', function incoming (message) {
+      const data = JSON.parse(message)
+      if (data.type === 'start') {
+        ws.send(JSON.stringify({ id: '1', type: 'complete' }))
+      }
+    })
+
+    ws.send(JSON.stringify({ id: undefined, type: 'connection_ack' }))
+  })
+
+  const client = new SubscriptionClient('ws://localhost:8889', {
+    reconnect: true,
+    maxReconnectAttempts: 10,
+    serviceName: 'test-service'
+  })
+
+  client.createSubscription('query', {}, (data) => {
+    t.deepEqual(data, {
+      topic: 'test-service_1',
+      payload: null
+    })
+    client.close()
+    server.close()
+    t.end()
+  })
+})
+
+test('subscription client tries to reconnect when server closes', (t) => {
+  let server = new WS.Server({ port: 8890 })
+  server.on('connection', function connection (ws) {
+    ws.send(JSON.stringify({ id: undefined, type: 'connection_ack' }))
+  })
+
+  const client = new SubscriptionClient('ws://localhost:8890', {
+    reconnect: true,
+    maxReconnectAttempts: 10,
+    serviceName: 'test-service'
+  })
+  server.close()
+
+  client.createSubscription('query', {}, (data) => {
+    t.deepEqual(data, {
+      topic: 'test-service_1',
+      payload: null
+    })
+    client.close()
+    server.close()
+    t.end()
+  })
+
+  setTimeout(() => {
+    server = new WS.Server({ port: 8890 })
+    server.on('connection', function connection (ws) {
+      ws.on('message', function incoming (message) {
+        const data = JSON.parse(message)
+        if (data.type === 'start') {
+          ws.send(JSON.stringify({ id: '1', type: 'complete' }))
+        }
+      })
+
+      ws.send(JSON.stringify({ id: undefined, type: 'connection_ack' }))
+    })
+  }, 2000)
+})
+
+test('subscription client stops trying reconnecting after maxReconnectAttempts', (t) => {
+  let server = new WS.Server({ port: 8891 })
+  server.on('connection', function connection (ws) {
+    ws.send(JSON.stringify({ id: undefined, type: 'connection_ack' }))
+  })
+
+  const client = new SubscriptionClient('ws://localhost:8891', {
+    reconnect: true,
+    maxReconnectAttempts: 2,
+    serviceName: 'test-service'
+  })
+  server.close()
+
+  client.createSubscription('query', {}, (data) => {
+    t.deepEqual(data, {
+      topic: 'test-service_1',
+      payload: null
+    })
+  })
+
+  setTimeout(() => {
+    server = new WS.Server({ port: 8891 })
+    server.on('connection', function connection (ws) {
+      throw new Error('Client connected when it should not!')
+    })
+    client.close()
+    server.close()
+    t.end()
+  }, 2000)
```

Comment:
Could we try to avoid `setTimeout` and use lolex or similar to fake the timers? This makes tests brittle